[hot_reload] Add instance fields

lambdageek opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/63643

Remaining work:
- [x] get/set fields in the debugger
- [x] Add instance properties
- [x] verify `ves_icall_property_info_get_default_value`works with added props
- [x] Add instance events (need added skeletons)
- [x] `mono_class_get_field_token`
- [x] `mono_class_get_event_token`
- [x] `mono_class_get_property_token`
- [x] `mono_field_get_rva`
- [x] `init_weak_fields_inner`
- [x] `ves_icall_RuntimeFieldInfo_GetFieldOffset`
- [x] `ves_icall_System_RuntimeFieldHandle_GetValueDirect`/`SetValueDirect`
- [x] audit callers of `mono_handle_unsafe_field_addr`
- [x] `mono_metadata_field_info_full`
- [x] `mono_metadata_get_marshal_info`, possibly

lambdageek committed:
[hot_reload] Add new AddInstanceField test

lambdageek committed:
WIP builds doesn't link; need impl for mono_metadata_update_ldflda

lambdageek committed:
WIP builds

lambdageek committed:
WIP test runs, fails on string load/stores

lambdageek committed:
Simple test works

lambdageek committed:
Add AddInstanceFieldToExistingType capability to runtime

lambdageek committed:
fix comment; remove unused var

lambdageek committed:
Add instance field: reflection and nested struct

lambdageek committed:
make TypedReference work with added fields

lambdageek committed:
Add FieldInfo.SetValue testcase

lambdageek committed:
make FieldInfo.SetValue work

lambdageek committed:
fix whitespace

lambdageek committed:
more whitespace

lambdageek committed:
remove TODO

lambdageek commented:
@thaystg @BrzVlad could you take a look.  Vlad: I'm adding an interp opcode and some transform changes to support it

lambdageek commented:
(Some of the code is part of #76618 )

lambdageek commented:
/azp run runtime-wasm

lambdageek committed:
fix Windows build error

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -7322,7 +7322,20 @@ MINT_IN_CASE(MINT_BRTRUE_I8_SP) ZEROP_SP(gint64, !=); MINT_IN_BREAK;
+			gpointer field_addr = mono_metadata_update_added_field_ldflda (inst, field_type, fielddef_token, error);
+			/* FIXME: think about pinning the FieldStore and adding a second opcode to
+			 * unpin it */
+			mono_gc_wbarrier_generic_store_internal (dest, field_addr);
```
This wbarrier is unnecessary. You can write the result like this: `LOCAL_VAR (ip [1], gpointer) = field_addr;`

lambdageek committed:
review feedback

lambdageek committed:
Implement LDFLDA opcode for added fields; add test

lambdageek committed:
fix build

lambdageek commented:
Found a few more items that need work before this is ready; updated the PR description

lambdageek commented:
Interesting. CoreCLR is crashing on something that I'm doing in the `AddInstanceField` test that I added in this PR :grin:

lambdageek commented:
CoreCLR failure is https://github.com/dotnet/runtime/issues/76702

lambdageek committed:
rm TODO

lambdageek committed:
Disable test on CoreCLR

lambdageek committed:
WIP start adding a wasm debugger test

lambdageek committed:
WIP XXX - semi-broken debugger test

lambdageek committed:
OK - test is failing in the debugger agent (see details)

* Assertion at
/Users/alklig/work/dotnet-runtime/runtime/src/mono/mono/component/debugger-agent.c:9829,
condition `!m_field_is_from_update (f)' not met

lambdageek committed:
make the debugger test a bit more interesting

lambdageek committed:
implement a debugger test that gets/sets instance fields

lambdageek committed:
Clear the debugger type cache on EnC update

we only need to clear the type on which fields or methods were
added. but right now we just clear the whole cache

lambdageek committed:
[sdb] implement get/set field for added instance fields

lambdageek committed:
[handles] remove the field getter/setter macros

They had no callers, and now we don't do things this way - we assume
the enclosing object had been pinned by a stack reference.

lambdageek committed:
[metadata-update] comments on asserts

lambdageek committed:
implement mono_field_get_addr added field support

lambdageek committed:
add TODO for initializers for added literals

lambdageek committed:
remove last use of MONO_HANDLE_SET_FIELD_REF

lambdageek committed:
WeakAttribute is not picked up from added fields

lambdageek committed:
Add a test for an array with RVA

lambdageek committed:
Impl mono_metadata_field_info_full

lambdageek committed:
mono_class_get_field_token

lambdageek committed:
mono_field_get_rva

lambdageek committed:
Add SetValueDirect/GetValueDirect test (failing)

lambdageek committed:
Implement GetValueDirect - hot reload test passes now

lambdageek committed:
fix init

lambdageek committed:
Added test for auto property; crashes in ves_icall_RuntimePropertyInfo_get_property_info

lambdageek committed:
fix rebase whitespace

lambdageek committed:
WIP checkpoint add_props

Note: we're not doing anything with the new PropertyMap row

Note2: we get MethodSemantics rows for properties that got updated,
and we're currently ignoring them.  Need to check that this is
reasonable. (A test would be to use reflection to grab a getter or
setter whose method body was changed and then try and invoke it and
verify that we're calling the correct method.)

lambdageek committed:
fixup comment - repeated MethodSemantics rows can happen

lambdageek committed:
added properties iteration

lambdageek committed:
protect callers of mono_class_get_property_token

lambdageek committed:
fix build warnings

lambdageek committed:
fix dynamic component builds

lambdageek committed:
make mono_class_get_property_token work for added props

lambdageek committed:
add test for adding instance event

lambdageek committed:
basic event reflection works

lambdageek committed:
Fire the new event, too

lambdageek committed:
make reflection MetadataToken work

lambdageek committed:
fix dynamic component build

lambdageek commented:
@thaystg @fanyang-mono This PR is essentially ready (I want to add a few more test cases for properties and events, but it's working locally, so I don't expect any other changes).  Could you please take another look.

thaystg commented on src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs:
```diff
@@ -561,5 +561,47 @@ public async Task DebugHotReloadMethod_AddingNewMethodWithoutAnyOtherChange_With
+            await SendCommandAndCheck (JObject.FromObject(new { }), "Debugger.resume", script_loc: null, line: -1, column: -1, function_name: null,
+                                       locals_fn: async (locals) => {
+                                           await CheckObject(locals, "c", "ApplyUpdateReferencedAssembly.AddInstanceFields.C");
+#if true
```
Why do we need this `#if true`?

lambdageek replied:
sorry. removed.

lambdageek committed:
remove unused ifdefs and fix whitespace
